sqldelight-runtime/src/test/java/com/squareup/sqldelight/TransactorTest.kt

```diff
@@ -0,0 +1,151 @@
+package com.squareup.sqldelight
+
+import com.google.common.truth.Truth.assertThat
+import com.squareup.sqldelight.db.SqlDatabaseConnection
+import com.squareup.sqldelight.sqlite.jdbc.SqliteJdbcOpenHelper
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import java.util.concurrent.atomic.AtomicInteger
+
+class TransactorTest {
+  private lateinit var database: Transactor
+  private lateinit var connection: SqlDatabaseConnection
+
+  @Before fun setup() {
+    val databaseHelper = SqliteJdbcOpenHelper()
+    database = object : Transactor(databaseHelper) {}
+    connection = databaseHelper.getConnection()
+  }
+
+  @After fun teardown() {
+    database.close()
+  }
+
+  @Test fun `afterCommit runs after transaction commits`() {
+    val counter = AtomicInteger(0)
+    database.transaction {
+      afterCommit { counter.incrementAndGet() }
+      assertThat(counter.get()).isEqualTo(0)
+    }
+
+    assertThat(counter.get()).isEqualTo(1)
+  }
+
+  @Test fun `afterCommit does not run after transaction rollbacks`() {
+    val counter = AtomicInteger(0)
+    database.transaction {
+      afterCommit { counter.incrementAndGet() }
+      assertThat(counter.get()).isEqualTo(0)
+      rollback()
+    }
+
+    assertThat(counter.get()).isEqualTo(0)
+  }
+
+  @Test fun `afterCommit runs immediately with no transaction`() {
+    val counter = AtomicInteger(0)
+    database.afterCommit { counter.incrementAndGet() }
+    assertThat(counter.get()).isEqualTo(1)
+  }
+
+  @Test fun `afterCommit runs after parent transaction commits`() {
+    val counter = AtomicInteger(0)
+    database.transaction {
+      afterCommit { counter.incrementAndGet() }
+      assertThat(counter.get()).isEqualTo(0)
+
+      transaction {
+        afterCommit { counter.incrementAndGet() }
+        assertThat(counter.get()).isEqualTo(0)
+      }
+
+      assertThat(counter.get()).isEqualTo(0)
+    }
+
+    assertThat(counter.get()).isEqualTo(2)
+  }
+
+  @Test fun `afterCommit does not run in nested transaction when parent rolls back`() {
+    val counter = AtomicInteger(0)
+    database.transaction {
+      afterCommit { counter.incrementAndGet() }
+      assertThat(counter.get()).isEqualTo(0)
+
+      transaction {
+        afterCommit { counter.incrementAndGet() }
+      }
+
+      rollback()
+    }
+
+    assertThat(counter.get()).isEqualTo(0)
+  }
+
+  @Test fun `afterCommit does not run in nested transaction when nested rolls back`() {
+    val counter = AtomicInteger(0)
+    database.transaction {
+      afterCommit { counter.incrementAndGet() }
+      assertThat(counter.get()).isEqualTo(0)
+
+      transaction {
+        afterCommit { counter.incrementAndGet() }
+        rollback()
+      }
+
+      throw AssertionError()
+    }
+
+    assertThat(counter.get()).isEqualTo(0)
+  }
+
+  @Test fun `afterRollback no-ops if there is no transaction`() {
```

Comment:
I’m sketched out by this. I think this should throw! Or perhaps be a protected method on `Transaction`